### PR TITLE
fix(evaluation):add request and limit for step

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Modules
- [ ] UI
- [x] Controller
- [ ] Agent
- [ ] Client
- [x] Python-SDK
- [ ] Others

## Checklist
- [ ] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/api/_impl/job.py [hunk 1]
```diff
@@ -1,30 +1,39 @@
...
 
     def decorator(func: t.Any) -> t.Any:
         if Parser.is_parse_stage():
-            cls, delim, func_name = func.__qualname__.rpartition(".")
+            cls, _, func_name = func.__qualname__.rpartition(".")
+            for _name, _resource in _resources.items():
+                if isinstance(_resource, numbers.Number):
+                    _resources[_name] = {"request": _resource, "limit": _resource}
+                elif isinstance(_resource, dict):
+                    if not all(n in resource_names for n, _ in _resource.items()):
```
Comment (tianweidut) [inline client/starwhale/api/_impl/job.py:33]: `n in resource_names for n, _ in _resource.items()` --> `n in resource_names for n in _resource` ?

File: client/starwhale/api/_impl/job.py [hunk 2]
```diff
@@ -1,30 +1,39 @@
...
-    ]
+    _resources = resources or {}
     _needs = needs or []
 
     def decorator(func: t.Any) -> t.Any:
         if Parser.is_parse_stage():
-            cls, delim, func_name = func.__qualname__.rpartition(".")
+            cls, _, func_name = func.__qualname__.rpartition(".")
+            for _name, _resource in _resources.items():
+                if isinstance(_resource, numbers.Number):
```
Comment (tianweidut) [inline client/starwhale/api/_impl/job.py:30]: add more validators?
cpu and memory accept int and float types.
gpu only accepts int type.

Comment (goldenxinxing) [reply client/starwhale/api/_impl/job.py]: done

File: client/starwhale/api/_impl/job.py (hunk 1 above)
Comment (goldenxinxing) [reply client/starwhale/api/_impl/job.py]: done

File: client/starwhale/api/_impl/job.py [hunk 3]
```diff
@@ -1,30 +1,59 @@
...
+from starwhale.consts import DEFAULT_EVALUATION_JOB_NAME
 from starwhale.core.job import dag
 from starwhale.utils.fs import ensure_file
 from starwhale.utils.load import load_module
 
+resource_names = {"cpu": numbers.Number, "gpu": int, "memory": numbers.Number}
+attribute_names = ["request", "limit"]
+
+
+def valid_resource(resources: t.Dict[str, t.Any]) -> None:
```
Comment (tianweidut) [inline client/starwhale/api/_impl/job.py:17]: `valid_resource` -> `do_resource_validate` ?

Comment (goldenxinxing) [reply client/starwhale/api/_impl/job.py]: done

File: server/controller/src/main/java/ai/starwhale/mlops/domain/runtime/RuntimeResource.java [hunk 4]
```diff
@@ -27,14 +27,6 @@
...
-    String type;
-    Float num;
-
-    public RuntimeResource(String str) {
-        String[] res = str.split("=");
-        this.type = res[0];
-        this.num = Float.valueOf(res[1]);
-    }
-
+    Float request;
```
[the remark this answers is not in the document]
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/domain/runtime/RuntimeResource.java:31]: recoverd

**Review by anda-ren — APPROVED** [review]
lgtm

**Pull request merged**